src/job_cache/daemon_cache.cpp

```diff
@@ -0,0 +1,1324 @@
+    exit(EXIT_FAILURE);
+  }
+
+  int pid = fork();
+
+  // error forking
+  if (pid < 0) {
+    perror("Failed to fork eviction process");
+    exit(EXIT_FAILURE);
+  }
+
+  // child
+  if (pid == 0) {
+    if (dup2(stdinPipe[read_side], STDIN_FILENO) == -1) {
+      perror("Failed to dup2 stdin pipe for eviction process");
+      exit(EXIT_FAILURE);
+    }
+
+    if (dup2(stdoutPipe[write_side], STDOUT_FILENO) == -1) {
+      perror("Failed to dup2 stdout pipe for eviction process");
+      exit(EXIT_FAILURE);
+    }
+
+    close(stdinPipe[read_side]);
+    close(stdinPipe[write_side]);
+    close(stdoutPipe[read_side]);
+    close(stdoutPipe[write_side]);
+
+    // Finally enter the eviction loop, if it exits cleanly
+    // go ahead and exit with its result.
+    int result =
+        eviction_loop(dir, std::make_unique<LRUEvictionPolicy>(max_cache_size, low_cache_size));
+    exit(result);
+  }
+
+  // parent
+  if (pid > 0) {
+    close(stdinPipe[read_side]);
+    close(stdoutPipe[write_side]);
+
+    evict_pid = pid;
+    evict_stdin = stdinPipe[write_side];
+    evict_stdout = stdoutPipe[read_side];
+  }
+}
+
+void DaemonCache::reap_evict_loop() {
+  close(evict_stdin);
+  close(evict_stdout);
+  waitpid(evict_pid, nullptr, 0);
+}
+
+// This has to be here because the destructor code for std::unique_ptr<CacheDbImpl>
+// requires that the CacheDbImpl be complete.
+DaemonCache::~DaemonCache() { reap_evict_loop(); }
+
+void DaemonCache::handle_new_client() {
+  // Accept the new client socket. Because of needing to perform
+  // multiple read's per client IO event, we have to make the client
+  // socket use non-blocking reads. This allows us to consume multiple
+  // read calls without blocking the other clients if one of them waits.
+  int accept_fd = accept4(listen_socket_fd, nullptr, nullptr, SOCK_CLOEXEC | SOCK_NONBLOCK);
+  if (accept_fd == -1) {
+    log_fatal("accept(%s): %s", key.c_str(), strerror(errno));
+  }
+
+  poll.add(accept_fd);
+  message_parsers.insert({accept_fd, MessageParser(accept_fd)});
+  log_info("new client connected: %d", accept_fd);
+}
+
+void DaemonCache::handle_msg(int client_fd) {
+  // In case multiple read events have been enqueued since the
+  // last epoll_wait, we have to perform all the reads that
+  // have been enqueued.
+  MessageParserState state;
+  std::vector<std::string> msgs;
+
+  auto it = message_parsers.find(client_fd);
+  if (it == message_parsers.end()) {
+    log_fatal("unreachable: message_parsers out of sync with poll. client_fd = %d", client_fd);
+  }
+
+  state = it->second.read_messages(msgs);
+
+  for (const auto &msg : msgs) {
+    // TODO: parse and dispatch message event
+    log_info("msg: %s", msg.c_str());
+
+    JAST json;
+    std::stringstream parseErrors;
+    if (!JAST::parse(msg, parseErrors, json)) {
+      log_fatal("DaemonCache::handle_msg(): failed to parse client request");
+    }
+
+    if (json.get("method").value == "cache/read") {
```

Comment:
this chunk of code should probably be moved to a "handle method call" function

Reply:
Should we make a hash table that maps method names to lambdas?

Reply:
I think that's silly on second though....also what the hell I worked for 10 hours today?!?

Reply:
I don't think we need it just yet, but this is actually what we do for the LSP

> also what the hell I worked for 10 hours today?!?
yeah we've been driving hard this week. Hopefully we get something to show for it soon